modules/service/src/main/scala/lucuma/odb/graphql/client/Itc.scala

```diff
@@ -0,0 +1,257 @@
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.graphql.client
+
+import cats.Order
+import cats.Order.catsKernelOrderingForOrder
+import cats.data.EitherNel
+import cats.data.NonEmptyList
+import cats.data.NonEmptySet
+import cats.effect.MonadCancel
+import cats.effect.Resource
+import cats.syntax.applicative.*
+import cats.syntax.bifunctor.*
+import cats.syntax.either.*
+import cats.syntax.flatMap.*
+import cats.syntax.functor.*
+import cats.syntax.option.*
+import cats.syntax.order.*
+import cats.syntax.traverse.*
+import eu.timepit.refined.types.numeric.NonNegInt
+import eu.timepit.refined.types.numeric.PosBigDecimal
+import lucuma.core.data.Zipper
+import lucuma.core.model.NonNegDuration
+import lucuma.core.model.Observation
+import lucuma.core.model.Program
+import lucuma.core.model.Target
+import lucuma.itc.client.ItcClient
+import lucuma.itc.client.ItcResult
+import lucuma.itc.client.SpectroscopyModeInput
+import lucuma.odb.service.ItcInputService
+import lucuma.odb.util.NonEmptyListExtensions.*
+
+/**
+ * The ITC client combines the input parameter lookup and the actual remote
+ * call to the ITC service in a single method.
+ */
+sealed trait Itc[F[_]] {
+  import Itc.ResultSet
+
+  def queryOne(
+    programId:     Program.Id,
+    observationId: Observation.Id,
+    useCache:      Boolean
+  ): F[Option[ResultSet]]
+
+  def queryAll(
+    programId:      Program.Id,
+    observationIds: List[Observation.Id],
+    useCache:       Boolean
+  ): F[List[ResultSet]]
+
+}
+
+
+object Itc {
+
+  opaque type Param = String
+
+  extension (p: Param) {
+    def stringValue: String =
+      p
+  }
+
+  /**
+   * Result of an ITC lookup for a single target of an observation, ignoring
+   * any other targets that an observation might have.
+   */
+  sealed trait Result extends Product with Serializable {
+
+    val missing: Option[Result.Missing] =
+      fold(_.some, _ => none, _ => none)
+
+    val serviceError: Option[Result.ServiceError] =
+      fold(_ => none, _.some, _ => none)
+
+    val success: Option[Result.Success] =
+      fold(_ => none, _ => none, _.some)
+
+    def fold[A](
+      mf: Result.Missing      => A,
+      ef: Result.ServiceError => A,
+      sf: Result.Success      => A
+    ): A =
+      this match {
+        case m @ Result.Missing(_, _)          => mf(m)
+        case e @ Result.ServiceError(_, _, _)  => ef(e)
+        case s @ Result.Success(_, _, _, _, _) => sf(s)
+      }
+
+  }
+
+  object Result {
+
+    /**
+     * One or more required parameters are missing, preventing the ITC service
+     * from being consulted.
+     */
+    final case class Missing(
+      targetId: Option[Target.Id],
+      params:   NonEmptySet[Param]
+    ) extends Result
+
+    def missing(
+      targetId: Option[Target.Id],
+      params:   NonEmptySet[Param]
+    ): Result =
+      Missing(targetId, params)
```

Comment:
If you use Scala 3 `enum` you don't need these widened constructors (but there may be a reason this data type won't work with `enum`).

Reply:
Thanks, that hadn't occurred to me but it makes it better.